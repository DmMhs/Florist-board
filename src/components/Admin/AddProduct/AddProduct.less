.AddProduct {
  width: 50%;
  border: 1px solid #eee;
  border-radius: 7px;
  padding: 1.5rem;
  display: flex;
  flex-direction: column;
  align-items: center;
  margin: 2rem auto;
  .form-control {
    display: flex;
    align-items: center;
    flex-wrap: wrap;
    width: 100%;
    input,
    select {
      margin-bottom: 1rem;
    }
    input {
      border: 1px solid #ccc;
      padding: 7px;
    }
    select {
      border: 1px solid #ccc;
    }
    textarea {
      width: 50%;
    }
    label {
      text-align: left;
      margin-bottom: 1rem;
      width: 100%;
    }
    &.product-images {
      .input-wrapper {
        width: 100%;
        display: flex;
      }
    }
    .price {
      select {
        margin-left: 1rem;
      }
    }
  }
  button {
    margin-top: 1rem;
    background-color: #eee;
    color: #333333;
    border: none;
    cursor: pointer;
    padding: 8px;
    font-size: 1.2rem;
    &:hover {
      background-color: #ccc;
    }
  }
}
