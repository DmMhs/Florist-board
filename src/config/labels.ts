const labels = {
  brand: 'FLORIST.UA',
  adress: 'MARKET SQUARE, LVIV, UA',
  postCode: '346346',
  phone: '+123-45-678',
  navigation: {
    home: 'HOME',
    shop: 'SHOP',
    gallery: 'GALLERY',
    contacts: 'CONTACTS',
    cart: 'CART',
    account: {
      main: 'MY ACCOUNT',
      menu: {
        signIn: 'SIGN IN',
        signUp: 'SIGN UP'
      }
    }
  }
};

export default labels;
