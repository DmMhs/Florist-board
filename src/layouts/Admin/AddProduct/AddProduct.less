.AddProduct, .EditProduct {
  .form-control {
    display: flex;
    align-items: center;
    flex-wrap: wrap;
    width: 100%;
    input,
    select {
      margin-bottom: 1rem;
    }
    input {
      border: 1px solid #ccc;
      padding: 7px;
    }
    select {
      border: 1px solid #ccc;
    }
    textarea {
      width: 50%;
    }
    label {
      text-align: left;
      margin-bottom: 1rem;
      width: 100%;
    }
    &.product-images {
      .input-wrapper {
        width: 100%;
        display: flex;
      }
    }
    .price {
      select {
        margin-left: 1rem;
      }
    }
  }
}
