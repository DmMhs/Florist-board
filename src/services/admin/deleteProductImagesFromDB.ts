import { productsRef } from '../../firebase';

export const deleteProductImagesFromDB = (id: string) => {
  return productsRef
    .child(id)
    .child('images')
    .remove()
    .catch(err => console.log(err));
};
