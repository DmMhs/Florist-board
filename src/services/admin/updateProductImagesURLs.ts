import { productsRef } from '../../firebase';

export const updateProductImagesURLs = (
  productKey: string,
  imagesURLs: string[]
) => {
  return productsRef
    .child(productKey)
    .update({
      images: imagesURLs
    })
    .catch(err => {
      console.log(err);
    });
};
