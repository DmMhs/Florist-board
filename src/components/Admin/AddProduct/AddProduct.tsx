import React, { Component } from 'react';

import './AddProduct.less';
import { AppContext } from '../../../AppContext';

interface AddProductProps {
  editModeEnabled?: boolean;
}
interface AddProductState {
  available: boolean;
  title: string;
  title_uk: string;
  images: File[];
  price: number;
  currency: string;
  description: string;
  description_uk: string;
}

class AddProduct extends Component<AddProductProps, AddProductState> {
  constructor(props: AddProductProps) {
    super(props);
    this.state = {
      available: true,
      title: '',
      title_uk: '',
      images: [],
      price: 0,
      currency: 'usd',
      description: '',
      description_uk: ''
    };
  }

  private descriptionChangedHandler = (
    event: React.ChangeEvent<HTMLTextAreaElement>
  ) => {
    this.setState({
      description: event.target.value
    });
  };
  private descriptionUkChangedHandler = (
    event: React.ChangeEvent<HTMLTextAreaElement>
  ) => {
    this.setState({
      description_uk: event.target.value
    });
  };

  private currencyChangedHandler = (
    event: React.ChangeEvent<HTMLSelectElement>
  ) => {
    this.setState({
      currency: event.target.value
    });
  };

  private priceInputChangedHandler = (
    event: React.ChangeEvent<HTMLInputElement>
  ) => {
    this.setState({
      price: +event.target.value
    });
  };

  private imgInputChangedHandler = (
    position: number,
    event: React.ChangeEvent<HTMLInputElement>
  ) => {
    const updatedImages: File[] = [...this.state.images];
    console.log(event.target.files);
    // updatedImages[position] = event.target.files as any;
    // this.setState({
    //   images: updatedImages
    // });
  };

  private titleInputChangedHandler = (
    event: React.ChangeEvent<HTMLInputElement>
  ) => {
    this.setState({
      title: event.target.value
    });
  };

  private titleUkInputChangedHandler = (
    event: React.ChangeEvent<HTMLInputElement>
  ) => {
    this.setState({
      title_uk: event.target.value
    });
  };

  private availableChangedHandler = (
    event: React.ChangeEvent<HTMLSelectElement>
  ) => {
    this.setState({
      available: event.target.value === 'true'
    });
  };

  private formSubmitHandler = (event: React.FormEvent<HTMLFormElement>) => {
    event.preventDefault();
    this.setState({
      available: true,
      title: '',
      title_uk: '',
      images: [],
      price: 0,
      currency: 'usd',
      description: '',
      description_uk: ''
    });
  };

  public render() {
    console.log(this.state);
    return (
      <AppContext.Consumer>
        {value =>
          value && (
            <form onSubmit={this.formSubmitHandler} className="AddProduct">
              <div className="form-control">
                <label>Product available:</label>
                <br />
                <select onChange={this.availableChangedHandler}>
                  <option value="true">yes</option>
                  <option value="false">no</option>
                </select>
              </div>
              <div className="form-control">
                <label>Product title:</label>
                <br />
                <input
                  type="text"
                  onChange={this.titleInputChangedHandler}
                  required
                  value={this.state.title as string}
                />
                <label>Product title uk: </label>
                <br />
                <input
                  type="text"
                  onChange={this.titleUkInputChangedHandler}
                  required
                  value={this.state.title_uk as string}
                />
              </div>
              <div className="form-control product-images">
                <label>Product images:</label>
                <br />
                <div className="input-wrapper">
                  <input
                    type="file"
                    onChange={this.imgInputChangedHandler.bind(this, 0)}
                  />
                </div>
                <div className="input-wrapper">
                  <input
                    type="file"
                    onChange={this.imgInputChangedHandler.bind(this, 1)}
                  />
                </div>
                <div className="input-wrapper">
                  <input
                    type="file"
                    onChange={this.imgInputChangedHandler.bind(this, 2)}
                  />
                </div>
              </div>
              <div className="form-control">
                <label>Product price:</label>
                <br />
                <div className="price">
                  <input
                    type="number"
                    required
                    onChange={this.priceInputChangedHandler}
                    value={this.state.price as number}
                  />
                  <select onChange={this.currencyChangedHandler}>
                    <option value="usd">USD</option>
                    <option value="eur">EUR</option>
                    <option value="uah">UAH</option>
                  </select>
                </div>
              </div>
              <div className="form-control">
                <label>Product description:</label>
                <br />
                <textarea
                  required
                  placeholder="Add product description"
                  onChange={this.descriptionChangedHandler}
                  value={this.state.description as string}
                />
                <label>Product description uk:</label>
                <br />
                <textarea
                  required
                  placeholder="Add product description"
                  onChange={this.descriptionUkChangedHandler}
                  value={this.state.description_uk as string}
                />
              </div>
              <button type="submit">SUBMIT</button>
              {this.state.images}
            </form>
          )
        }
      </AppContext.Consumer>
    );
  }
}

export default AddProduct;
