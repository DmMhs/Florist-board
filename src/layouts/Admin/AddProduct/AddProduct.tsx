import React, { Component } from 'react';

import './AddProduct.less';
import { AppContext } from '../../../AppContext';
import { storageRef, productsRef } from '../../../firebase';
import { withRouter, RouteComponentProps } from 'react-router';
import { deleteProductImages } from '../../../services/deleteProductImages';

interface AddProductProps {
  editModeEnabled?: boolean;
}
interface AddProductState {
  available: boolean;
  title: string;
  title_ua: string;
  images: File[];
  price: number;
  currency: string;
  description: string;
  description_ua: string;
  productId?: string;
  imagesFolderName?: string;
}

class AddProduct extends Component<
  RouteComponentProps<{}> & AddProductProps,
  AddProductState
> {
  public static getDerivedStateFromProps(
    props: RouteComponentProps<{}> & AddProductProps,
    state: AddProductState
  ) {
    const url = props.location.search;
    const params = new URLSearchParams(url);
    const extractedId = params.get('id');
    if (extractedId !== null && extractedId !== undefined) {
      return {
        productId: extractedId
      };
    } else {
      return {};
    }
  }
  private editAvailableRef: React.RefObject<HTMLSelectElement>;

  constructor(props: RouteComponentProps<{}> & AddProductProps) {
    super(props);
    this.state = {
      available: true,
      title: '',
      title_ua: '',
      images: [],
      price: 0,
      currency: 'usd',
      description: '',
      description_ua: ''
    };
    this.editAvailableRef = React.createRef();
  }

  public componentDidMount() {
    if (this.props.editModeEnabled === true) {
      productsRef
        .child(this.state.productId as string)
        .on('value', snapshot => {
          const {
            available,
            title,
            title_ua,
            price,
            currency,
            description,
            description_ua
          } = snapshot!.val();
          this.setState({
            available,
            title,
            title_ua,
            price,
            currency,
            description,
            description_ua,
            imagesFolderName: title.toLowerCase()
          });
        });
    }
  }

  private descriptionChangedHandler = (
    event: React.ChangeEvent<HTMLTextAreaElement>
  ) => {
    this.setState({
      description: event.target.value
    });
  };

  private descriptionUAChangedHandler = (
    event: React.ChangeEvent<HTMLTextAreaElement>
  ) => {
    this.setState({
      description_ua: event.target.value
    });
  };

  private currencyChangedHandler = (
    event: React.ChangeEvent<HTMLSelectElement>
  ) => {
    this.setState({
      currency: event.target.value
    });
  };

  private priceInputChangedHandler = (
    event: React.ChangeEvent<HTMLInputElement>
  ) => {
    this.setState({
      price: +event.target.value
    });
  };

  private imgInputChangedHandler = (
    position: number,
    event: React.ChangeEvent<HTMLInputElement>
  ) => {
    const updatedImages: File[] = [...this.state.images];
    updatedImages[position] = event.target.files as any;
    this.setState({
      images: updatedImages
    });
  };

  private titleInputChangedHandler = (
    event: React.ChangeEvent<HTMLInputElement>
  ) => {
    this.setState({
      title: event.target.value
    });
  };

  private titleUAInputChangedHandler = (
    event: React.ChangeEvent<HTMLInputElement>
  ) => {
    this.setState({
      title_ua: event.target.value
    });
  };

  private availableChangedHandler = (
    event: React.ChangeEvent<HTMLSelectElement>
  ) => {
    this.setState({
      available: event.target.value === 'true'
    });
  };

  private formSubmitHandler = async (
    event: React.FormEvent<HTMLFormElement>
  ) => {
    event.preventDefault();

    const {
      available,
      images,
      title,
      title_ua,
      price,
      currency,
      description,
      description_ua
    } = this.state;

    const formattedFolderName = title.toLowerCase();
    const imagesRef = storageRef
      .child('products-images')
      .child(formattedFolderName);
    const imageURLs: string[] = [];
    let productKey: string = '';

    const newProduct = {
      title,
      title_ua,
      price,
      available,
      description,
      description_ua,
      currency,
      images
    };

    if (this.props.editModeEnabled === true) {
      productKey = this.state.productId as string;
      await deleteProductImages(
        this.state.productId as string,
        this.state.imagesFolderName as string
      );
      await productsRef
        .child(this.state.productId as string)
        .child('images')
        .remove()
        .catch(err => console.log(err));

      await productsRef
        .child(this.state.productId as string)
        .set({ ...newProduct })
        .catch(err => console.log(err));

      await Promise.all(
        images.map(async (image: File) => {
          const file = (image as any)[0];
          const formattedFileName = (image as any)[0].name.split('.')[0];
          await storageRef
            .child('products-images')
            .child(newProduct.title.toLowerCase())
            .child(formattedFileName)
            .put(file)
            .catch(err => {
              console.log(err);
            });

          const imageURL = await storageRef
            .child('products-images')
            .child(newProduct.title.toLowerCase())
            .child(formattedFileName)
            .getDownloadURL()
            .catch(err => console.log(err));
          imageURLs.push(imageURL);
        })
      );

      productsRef
        .child(productKey)
        .update({
          images: imageURLs
        })
        .catch(err => {
          console.log(err);
        });
    } else {
      await productsRef
        .push(newProduct)
        .then(response => {
          productKey = response.key as string;
        })
        .catch(err => {
          console.log(err);
        });

      await Promise.all(
        images.map(async (image: File) => {
          const file = (image as any)[0];
          const formattedFileName = (image as any)[0].name.split('.')[0];
          await imagesRef
            .child(formattedFileName)
            .put(file)
            .catch(err => {
              console.log(err);
            });
          const imageURL = await imagesRef
            .child(formattedFileName)
            .getDownloadURL();
          imageURLs.push(imageURL);
        })
      );

      productsRef
        .child(productKey)
        .update({
          images: imageURLs
        })
        .catch(err => {
          console.log(err);
        });
    }
    this.setState({
      available: true,
      title: '',
      title_ua: '',
      images: [],
      price: 0,
      currency: 'usd',
      description: '',
      description_ua: ''
    });
  };

  public render() {
    const context = this.context;
    const labels = context.state.labels;
    const lang = context.state.lang;

    const labelsRoot = labels[lang].pages.admin.addProductForm;
    const submitBtnLabel = labels[lang].pages.admin.submitBtn;

    const editModeEnabled = this.props.editModeEnabled;
    let editProductForm: JSX.Element;

    if (editModeEnabled === true) {
      editProductForm = (
        <form onSubmit={this.formSubmitHandler} className="AddProduct form">
          <div className="form-control">
            <label>{labelsRoot.available.title}</label>
            <br />
            <select
              onChange={this.availableChangedHandler}
              ref={this.editAvailableRef}
              value={this.state.available.toString()}
              className="availableSelect"
            >
              <option value="true">{labelsRoot.available.option1}</option>
              <option value="false">{labelsRoot.available.option2}</option>
            </select>
          </div>
          <div className="form-control">
            <label>{labelsRoot.title}</label>
            <br />
            <input
              type="text"
              onChange={this.titleInputChangedHandler}
              required
              value={this.state.title as string}
              placeholder={this.state.title}
              className="titleInput"
            />
            <label>{labelsRoot.title_ua}</label>
            <br />
            <input
              type="text"
              onChange={this.titleUAInputChangedHandler}
              required
              value={this.state.title_ua as string}
              placeholder={this.state.title_ua}
              className="titleUAInput"
            />
          </div>
          <div className="form-control product-images">
            <label>{labelsRoot.images}</label>
            <br />
            <div className="input-wrapper">
              <input
                type="file"
                onChange={this.imgInputChangedHandler.bind(this, 0)}
                required
                className="fileInput"
              />
            </div>
            <div className="input-wrapper">
              <input
                type="file"
                onChange={this.imgInputChangedHandler.bind(this, 1)}
                required
                className="fileInput"
              />
            </div>
            <div className="input-wrapper">
              <input
                type="file"
                onChange={this.imgInputChangedHandler.bind(this, 2)}
                required
                className="fileInput"
              />
            </div>
          </div>
          <div className="form-control">
            <label>{labelsRoot.price}</label>
            <br />
            <div className="price">
              <input
                type="number"
                required
                onChange={this.priceInputChangedHandler}
                value={this.state.price as number}
                placeholder={this.state.price.toString()}
                className="priceInput"
              />
              <select
                onChange={this.currencyChangedHandler}
                value={this.state.currency}
                className="currencySelect"
              >
                <option value="usd">USD</option>
                <option value="eur">EUR</option>
                <option value="uah">UAH</option>
              </select>
            </div>
          </div>
          <div className="form-control">
            <label>{labelsRoot.description}</label>
            <br />
            <textarea
              required
              placeholder={this.state.description}
              onChange={this.descriptionChangedHandler}
              value={this.state.description as string}
              className="descriptionText"
            />
            <label>{labelsRoot.description_ua}</label>
            <br />
            <textarea
              required
              placeholder={this.state.description_ua}
              onChange={this.descriptionUAChangedHandler}
              value={this.state.description_ua as string}
              className="descriptionUAText"
            />
          </div>
          <button type="submit">{submitBtnLabel}</button>
        </form>
      );
    }

    const addProductForm = (
      <form onSubmit={this.formSubmitHandler} className="AddProduct form">
        <div className="form-control">
          <label>{labelsRoot.available.title}</label>
          <br />
          <select
            onChange={this.availableChangedHandler}
            className="availableSelect"
          >
            <option value="true">{labelsRoot.available.option1}</option>
            <option value="false">{labelsRoot.available.option2}</option>
          </select>
        </div>
        <div className="form-control">
          <label>{labelsRoot.title}</label>
          <br />
          <input
            type="text"
            onChange={this.titleInputChangedHandler}
            required
            value={this.state.title as string}
            className="titleInput"
          />
          <label>{labelsRoot.title_ua}</label>
          <br />
          <input
            type="text"
            onChange={this.titleUAInputChangedHandler}
            required
            value={this.state.title_ua as string}
            className="titleUAInput"
          />
        </div>
        <div className="form-control product-images">
          <label>{labelsRoot.images}</label>
          <br />
          <div className="input-wrapper">
            <input
              type="file"
              onChange={this.imgInputChangedHandler.bind(this, 0)}
              required
              className="fileInput"
            />
          </div>
          <div className="input-wrapper">
            <input
              type="file"
              onChange={this.imgInputChangedHandler.bind(this, 1)}
              required
              className="fileInput"
            />
          </div>
          <div className="input-wrapper">
            <input
              type="file"
              onChange={this.imgInputChangedHandler.bind(this, 2)}
              required
              className="fileInput"
            />
          </div>
        </div>
        <div className="form-control">
          <label>{labelsRoot.price}</label>
          <br />
          <div className="price">
            <input
              type="number"
              required
              onChange={this.priceInputChangedHandler}
              value={this.state.price as number}
              className="priceInput"
            />
            <select
              onChange={this.currencyChangedHandler}
              className="currencySelect"
            >
              <option value="usd">USD</option>
              <option value="eur">EUR</option>
              <option value="uah">UAH</option>
            </select>
          </div>
        </div>
        <div className="form-control">
          <label>{labelsRoot.description}</label>
          <br />
          <textarea
            required
            placeholder="Add product description"
            onChange={this.descriptionChangedHandler}
            value={this.state.description as string}
            className="descriptionText"
          />
          <label>{labelsRoot.description_ua}</label>
          <br />
          <textarea
            required
            placeholder="Add product description"
            onChange={this.descriptionUAChangedHandler}
            value={this.state.description_ua as string}
            className="descriptionUAText"
          />
        </div>
        <button type="submit">{submitBtnLabel}</button>
      </form>
    );

    return (
      <AppContext.Consumer>
        {value =>
          value && (editModeEnabled === true ? editProductForm : addProductForm)
        }
      </AppContext.Consumer>
    );
  }
}

AddProduct.contextType = AppContext;
export default withRouter<AddProductProps & RouteComponentProps<{}>>(
  AddProduct
);
