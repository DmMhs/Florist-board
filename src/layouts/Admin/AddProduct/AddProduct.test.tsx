import React from 'react';
import { mount } from 'enzyme';
import { BrowserRouter } from 'react-router-dom';

import AddProduct from './AddProduct';
import AppContextProvider from '../../../AppContext';
import labels from '../../../config/labels';
import * as deleteProductImagesFunction from '../../../services/admin/deleteProductImages';
import * as deleteProductImagesFromDBFunction from '../../../services/admin/deleteProductImagesFromDB';
import * as updateProductImagesURLsFunction from '../../../services/admin/updateProductImagesURLs';

describe('AddProduct works as expected', () => {
  it('AddProduct component matches a snapshot', () => {
    const wrapper = mount(
      <BrowserRouter>
        <AppContextProvider>
          <AddProduct />
        </AppContextProvider>
      </BrowserRouter>
    );
    const context = wrapper.find('AppContextProvider').instance();
    context.setState({
      lang: 'en',
      labels: labels,
      fetchInProgress: false,
      mobileMode: true,
      showNavigation: false,
      togglePosition: 'absolute'
    });
    wrapper.update();
    expect(wrapper).toMatchSnapshot();
  });

  it('descriptionChangedHandler & descriptionUAChangedHandler change the state', () => {
    const wrapper = mount(
      <BrowserRouter>
        <AppContextProvider>
          <AddProduct />
        </AppContextProvider>
      </BrowserRouter>
    );
    const context = wrapper.find('AppContextProvider').instance();
    context.setState({
      userLogin: 'testLogin',
      userId: 'testId',
      userToken: 'testToken',
      userRole: 'admin',
      userAuthenticated: true,
      authenticationMethod: undefined,
      lang: 'en',
      labels: labels,
      fetchInProgress: false,
      mobileMode: true,
      showNavigation: false,
      togglePosition: 'absolute'
    });
    wrapper.update();
    const instance = wrapper.find('AddProduct').instance();
    wrapper.find('.descriptionText').simulate('change', {
      target: { value: 'test descriptionChangedHandler' }
    });
    wrapper.find('.descriptionUAText').simulate('change', {
      target: { value: 'test descriptionUAChangedHandler' }
    });
    expect(instance.state.description).toEqual(
      'test descriptionChangedHandler'
    );
    expect(instance.state.description_ua).toEqual(
      'test descriptionUAChangedHandler'
    );
  });
  it('currencyChangedHandler changes the state', () => {
    const wrapper = mount(
      <BrowserRouter>
        <AppContextProvider>
          <AddProduct />
        </AppContextProvider>
      </BrowserRouter>
    );
    const context = wrapper.find('AppContextProvider').instance();
    context.setState({
      userLogin: 'testLogin',
      userId: 'testId',
      userToken: 'testToken',
      userRole: 'admin',
      userAuthenticated: true,
      authenticationMethod: undefined,
      lang: 'en',
      labels: labels,
      fetchInProgress: false,
      mobileMode: true,
      showNavigation: false,
      togglePosition: 'absolute'
    });
    wrapper.update();
    const instance = wrapper.find('AddProduct').instance();
    wrapper.find('.currencySelect').simulate('change', {
      target: { value: 'UAH' }
    });
    expect(instance.state.currency).toEqual('UAH');
  });
  it('priceInputChangedHandler changes the state', () => {
    const wrapper = mount(
      <BrowserRouter>
        <AppContextProvider>
          <AddProduct />
        </AppContextProvider>
      </BrowserRouter>
    );
    const context = wrapper.find('AppContextProvider').instance();
    context.setState({
      userLogin: 'testLogin',
      userId: 'testId',
      userToken: 'testToken',
      userRole: 'admin',
      userAuthenticated: true,
      authenticationMethod: undefined,
      lang: 'en',
      labels: labels,
      fetchInProgress: false,
      mobileMode: true,
      showNavigation: false,
      togglePosition: 'absolute'
    });
    wrapper.update();
    const instance = wrapper.find('AddProduct').instance();
    wrapper.find('.priceInput').simulate('change', {
      target: { value: 200 }
    });
    expect(instance.state.price).toEqual(200);
  });
  it('imgInputChangedHandler changes the state', () => {
    const wrapper = mount(
      <BrowserRouter>
        <AppContextProvider>
          <AddProduct />
        </AppContextProvider>
      </BrowserRouter>
    );
    const context = wrapper.find('AppContextProvider').instance();
    context.setState({
      userLogin: 'testLogin',
      userId: 'testId',
      userToken: 'testToken',
      userRole: 'admin',
      userAuthenticated: true,
      authenticationMethod: undefined,
      lang: 'en',
      labels: labels,
      fetchInProgress: false,
      mobileMode: true,
      showNavigation: false,
      togglePosition: 'absolute'
    });
    const file = 'dummy.image';
    wrapper.update();
    const instance = wrapper.find('AddProduct').instance();
    wrapper
      .find('.fileInput')
      .at(0)
      .simulate('change', {
        target: { files: file }
      });
    expect(instance.state.images).toEqual(['dummy.image']);
  });
  it('titleInputChangedHandler & titleUAInputChangedHandler change the state', () => {
    const wrapper = mount(
      <BrowserRouter>
        <AppContextProvider>
          <AddProduct />
        </AppContextProvider>
      </BrowserRouter>
    );
    const context = wrapper.find('AppContextProvider').instance();
    context.setState({
      userLogin: 'testLogin',
      userId: 'testId',
      userToken: 'testToken',
      userRole: 'admin',
      userAuthenticated: true,
      authenticationMethod: undefined,
      lang: 'en',
      labels: labels,
      fetchInProgress: false,
      mobileMode: true,
      showNavigation: false,
      togglePosition: 'absolute'
    });
    wrapper.update();
    const instance = wrapper.find('AddProduct').instance();
    wrapper.find('.titleInput').simulate('change', {
      target: { value: 'test titleInputChangedHandler' }
    });
    wrapper.find('.titleUAInput').simulate('change', {
      target: { value: 'test titleUAInputChangedHandler' }
    });
    expect(instance.state.title).toEqual('test titleInputChangedHandler');
    expect(instance.state.title_ua).toEqual('test titleUAInputChangedHandler');
  });
  it('availableChangedHandler changes the state', () => {
    const wrapper = mount(
      <BrowserRouter>
        <AppContextProvider>
          <AddProduct />
        </AppContextProvider>
      </BrowserRouter>
    );
    const context = wrapper.find('AppContextProvider').instance();
    context.setState({
      userLogin: 'testLogin',
      userId: 'testId',
      userToken: 'testToken',
      userRole: 'admin',
      userAuthenticated: true,
      authenticationMethod: undefined,
      lang: 'en',
      labels: labels,
      fetchInProgress: false,
      mobileMode: true,
      showNavigation: false,
      togglePosition: 'absolute'
    });
    wrapper.update();
    const instance = wrapper.find('AddProduct').instance();
    wrapper.find('.availableSelect').simulate('change', {
      target: { value: 'true' }
    });
    expect(instance.state.available).toEqual(true);
  });
  it('formSubmitHandler changes the state', () => {
    const wrapper = mount(
      <BrowserRouter>
        <AppContextProvider>
          <AddProduct />
        </AppContextProvider>
      </BrowserRouter>
    );
    const context = wrapper.find('AppContextProvider').instance();
    context.setState({
      userLogin: 'testLogin',
      userId: 'testId',
      userToken: 'testToken',
      userRole: 'admin',
      userAuthenticated: true,
      authenticationMethod: undefined,
      lang: 'en',
      labels: labels,
      fetchInProgress: false,
      mobileMode: true,
      showNavigation: false,
      togglePosition: 'absolute'
    });
    wrapper.update();
    const instance = wrapper.find('AddProduct').instance();
    instance.setState({
      available: true,
      title: 'test',
      title_ua: 'test',
      images: [],
      price: 30,
      currency: 'usd',
      description: 'test',
      description_ua: 'test'
    });
    wrapper.find('.form').simulate('submit');
    const asyncCheck = setImmediate(() => {
      wrapper.update();
      expect(instance.state).toEqual({
        available: true,
        title: '',
        title_ua: '',
        images: [],
        price: 0,
        currency: 'usd',
        description: '',
        description_ua: ''
      });
    });
    global.clearImmediate(asyncCheck);
  });
  it('editModeEnabled and submit envokes helper functions', async () => {
    const wrapper = mount(
      <BrowserRouter>
        <AppContextProvider>
          <AddProduct editModeEnabled={true} />
        </AppContextProvider>
      </BrowserRouter>
    );
    const context = wrapper.find('AppContextProvider').instance();
    context.setState({
      userLogin: 'testLogin',
      userId: 'testId',
      userToken: 'testToken',
      userRole: 'admin',
      userAuthenticated: true,
      authenticationMethod: undefined,
      lang: 'en',
      labels: labels,
      fetchInProgress: false,
      mobileMode: true,
      showNavigation: false,
      togglePosition: 'absolute'
    });
    wrapper.update();
    const instance = wrapper.find('AddProduct').instance();

    const p = Promise.resolve();

    const spyDeleteImages = jest
      .spyOn(deleteProductImagesFunction, 'deleteProductImages')
      .mockImplementation((id: string, folderName: string) => p);
    const spyDeleteImagesFromDB = jest
      .spyOn(deleteProductImagesFromDBFunction, 'deleteProductImagesFromDB')
      .mockImplementation((id: string) => p);
    // const spyUpdateImageURL = jest
    //   .spyOn(updateProductImagesURLsFunction, 'updateProductImagesURLs')
    //   .mockImplementation(() => p);
    instance.forceUpdate();
    wrapper.find('.form').simulate('submit');
    wrapper.update();
    await p;
    expect(spyDeleteImages).toHaveBeenCalled();
    expect(spyDeleteImagesFromDB).toHaveBeenCalled();
    // expect(spyUpdateImageURL).toHaveBeenCalled();
  });
});
