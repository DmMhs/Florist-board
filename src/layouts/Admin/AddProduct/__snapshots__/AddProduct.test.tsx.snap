// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddProduct works as expected AddProduct component matches a snapshot 1`] = `
<BrowserRouter>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "createHref": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <withRouter(AppContextProvider)>
      <Route>
        <AppContextProvider
          history={
            Object {
              "action": "POP",
              "block": [Function],
              "createHref": [Function],
              "go": [Function],
              "goBack": [Function],
              "goForward": [Function],
              "length": 1,
              "listen": [Function],
              "location": Object {
                "hash": "",
                "pathname": "/",
                "search": "",
                "state": undefined,
              },
              "push": [Function],
              "replace": [Function],
            }
          }
          location={
            Object {
              "hash": "",
              "pathname": "/",
              "search": "",
              "state": undefined,
            }
          }
          match={
            Object {
              "isExact": true,
              "params": Object {},
              "path": "/",
              "url": "/",
            }
          }
        >
          <withRouter(AddProduct)>
            <Route>
              <AddProduct
                history={
                  Object {
                    "action": "POP",
                    "block": [Function],
                    "createHref": [Function],
                    "go": [Function],
                    "goBack": [Function],
                    "goForward": [Function],
                    "length": 1,
                    "listen": [Function],
                    "location": Object {
                      "hash": "",
                      "pathname": "/",
                      "search": "",
                      "state": undefined,
                    },
                    "push": [Function],
                    "replace": [Function],
                  }
                }
                location={
                  Object {
                    "hash": "",
                    "pathname": "/",
                    "search": "",
                    "state": undefined,
                  }
                }
                match={
                  Object {
                    "isExact": true,
                    "params": Object {},
                    "path": "/",
                    "url": "/",
                  }
                }
              >
                <form
                  className="AddProduct form"
                  onSubmit={[Function]}
                >
                  <div
                    className="form-control"
                  >
                    <label>
                      Product available: 
                    </label>
                    <br />
                    <select
                      className="availableSelect"
                      onChange={[Function]}
                    >
                      <option
                        value="true"
                      >
                        yes
                      </option>
                      <option
                        value="false"
                      >
                        no
                      </option>
                    </select>
                  </div>
                  <div
                    className="form-control"
                  >
                    <label>
                      Product title (EN): 
                    </label>
                    <br />
                    <input
                      className="titleInput"
                      onChange={[Function]}
                      required={true}
                      type="text"
                      value=""
                    />
                    <label>
                      Product title (UA): 
                    </label>
                    <br />
                    <input
                      className="titleUAInput"
                      onChange={[Function]}
                      required={true}
                      type="text"
                      value=""
                    />
                  </div>
                  <div
                    className="form-control product-images"
                  >
                    <label>
                      Product images: 
                    </label>
                    <br />
                    <div
                      className="input-wrapper"
                    >
                      <input
                        className="fileInput"
                        onChange={[Function]}
                        required={true}
                        type="file"
                      />
                    </div>
                    <div
                      className="input-wrapper"
                    >
                      <input
                        className="fileInput"
                        onChange={[Function]}
                        required={true}
                        type="file"
                      />
                    </div>
                    <div
                      className="input-wrapper"
                    >
                      <input
                        className="fileInput"
                        onChange={[Function]}
                        required={true}
                        type="file"
                      />
                    </div>
                  </div>
                  <div
                    className="form-control"
                  >
                    <label>
                      Product price: 
                    </label>
                    <br />
                    <div
                      className="price"
                    >
                      <input
                        className="priceInput"
                        onChange={[Function]}
                        required={true}
                        type="number"
                        value={0}
                      />
                      <select
                        className="currencySelect"
                        onChange={[Function]}
                      >
                        <option
                          value="usd"
                        >
                          USD
                        </option>
                        <option
                          value="eur"
                        >
                          EUR
                        </option>
                        <option
                          value="uah"
                        >
                          UAH
                        </option>
                      </select>
                    </div>
                  </div>
                  <div
                    className="form-control"
                  >
                    <label>
                      Product description (EN): 
                    </label>
                    <br />
                    <textarea
                      className="descriptionText"
                      onChange={[Function]}
                      placeholder="Add product description"
                      required={true}
                      value=""
                    />
                    <label>
                      Product description (UA): 
                    </label>
                    <br />
                    <textarea
                      className="descriptionUAText"
                      onChange={[Function]}
                      placeholder="Add product description"
                      required={true}
                      value=""
                    />
                  </div>
                  <button
                    type="submit"
                  >
                    SUBMIT
                  </button>
                </form>
              </AddProduct>
            </Route>
          </withRouter(AddProduct)>
        </AppContextProvider>
      </Route>
    </withRouter(AppContextProvider)>
  </Router>
</BrowserRouter>
`;
